File: evaluator/evaluator.go
package evaluator

import (
	"fmt"
	"gorb/ast"
	"gorb/object"
	"gorb/token"
)

var (
	NULL  = &object.Null{}
	TRUE  = &object.Boolean{Value: true}
	FALSE = &object.Boolean{Value: false}
)

func Eval(node ast.Node, env *object.Environment) object.Object {
	switch node := node.(type) {
	case *ast.Program:
		return evalProgram(node, env)

	// statements
	case *ast.ReturnStatement:
		val := Eval(node.ReturnValue, env)
		if isError(val) {
			return val
		}
		return &object.ReturnValue{Value: val}

	case *ast.DeclarationStatement:
		val := Eval(node.Value, env)
		if isError(val) {
			return val
		}
		env.Set(node.Name.Value, val)

	case *ast.ExpressionStatement:
		return Eval(node.Expression, env)

	case *ast.BlockStatement:
		return evalBlockStatement(node, env)

	// expressions
	case *ast.IndexExpression:
		left := Eval(node.Left, env)
		if isError(left) {
			return left
		}
		index := Eval(node.Index, env)
		if isError(index) {
			return index
		}
		return evalIndexExpression(left, index)

	case *ast.CallExpression:
		fn := Eval(node.Function, env)
		if isError(fn) {
			return fn
		}
		args := evalExpressions(node.Arguments, env)
		if len(args) == 1 && isError(args[0]) {
			return args[0]
		}
		return applyFunction(fn, args)

	case *ast.IfExpression:
		return evalIfExpression(node, env)

	case *ast.UnaryExpression:
		right := Eval(node.Right, env)
		if isError(right) {
			return right
		}
		return evalUnaryExpression(node.Operator, right)

	case *ast.BinaryExpression:
		left := Eval(node.Left, env)
		if isError(left) {
			return left
		}
		right := Eval(node.Right, env)
		if isError(right) {
			return right
		}
		return evalBinaryExpression(node.Operator, left, right)

	// literals
	case *ast.FunctionLiteral:
		params := node.Parameters
		body := node.Body
		return &object.Function{Parameters: params, Body: body, Env: env}

	case *ast.Identifier:
		return evalIdentifier(node, env)

	case *ast.BooleanLiteral:
		return boolToBooleanObject(node.Value)

	case *ast.IntegerLiteral:
		return &object.Integer{Value: node.Value}

	case *ast.StringLiteral:
		return &object.String{Value: node.Value}

	case *ast.ArrayLiteral:
		elements := evalExpressions(node.Elements, env)
		if len(elements) == 1 && isError(elements[0]) {
			return elements[0]
		}
		return &object.Array{Elements: elements}

	case *ast.HashLiteral:
		return evalHashLiteral(node, env)

	}

	return nil
}

func evalProgram(program *ast.Program, env *object.Environment) object.Object {
	var result object.Object

	for _, stmt := range program.Statements {
		result = Eval(stmt, env)

		switch result := result.(type) {
		case *object.ReturnValue:
			return result.Value
		case *object.Error:
			return result
		}
	}

	return result
}

// statements

func evalBlockStatement(
	block *ast.BlockStatement,
	env *object.Environment,
) object.Object {
	var result object.Object

	for _, stmt := range block.Statements {
		result = Eval(stmt, env)

		if result != nil {
			rt := result.Type()
			if rt == object.RETURN_VALUE || rt == object.ERROR {
				return result
			}
		}
	}

	return result
}

// expressions

func evalExpressions(
	exps []ast.Expression,
	env *object.Environment,
) []object.Object {
	var result []object.Object

	for _, e := range exps {
		val := Eval(e, env)
		if isError(val) {
			return []object.Object{val}
		}
		result = append(result, val)
	}

	return result
}

func evalIndexExpression(left, index object.Object) object.Object {
	switch {
	case left.Type() == object.ARRAY && index.Type() == object.INTEGER:
		return evalArrayIndexExpression(left, index)
	case left.Type() == object.HASH:
		return evalHashIndexExpression(left, index)
	default:
		return newError("index operator not supported: %s", left.Type())
	}
}

func evalArrayIndexExpression(array, index object.Object) object.Object {
	arrayObject := array.(*object.Array)
	idx := index.(*object.Integer).Value
	max := int64(len(arrayObject.Elements) - 1)

	if idx < 0 || idx > max {
		return NULL
	}

	return arrayObject.Elements[idx]
}

func evalHashIndexExpression(hash, index object.Object) object.Object {
	hashObject := hash.(*object.Hash)

	key, ok := index.(object.Hashable)
	if !ok {
		return newError("unusable as hash key: %s", index.Type())
	}

	pair, ok := hashObject.Pairs[key.HashKey()]
	if !ok {
		return NULL
	}

	return pair.Value
}

func evalIfExpression(
	ie *ast.IfExpression,
	env *object.Environment,
) object.Object {
	condition := Eval(ie.Condition, env)
	if isError(condition) {
		return condition
	}

	if isTruthy(condition) {
		return Eval(ie.Consequence, env)
	} else if ie.Alternative != nil {
		return Eval(ie.Alternative, env)
	} else {
		return NULL
	}
}

func evalUnaryExpression(
	operator token.TokenType,
	right object.Object,
) object.Object {
	switch operator {
	case token.MINUS:
		return evalInverseExpression(right)
	case token.BANG:
		return evalNegateExpression(right)
	default:
		return newError("unknown operation: %s%s", operator, right.Type())
	}
}

func evalInverseExpression(right object.Object) object.Object {
	if right.Type() != object.INTEGER {
		return newError("unknown operation: -%s", right.Type())
	}

	val := right.(*object.Integer).Value
	return &object.Integer{Value: -val}
}

func evalNegateExpression(right object.Object) object.Object {
	switch right {
	case TRUE:
		return FALSE
	case FALSE:
		return TRUE
	case NULL:
		return TRUE
	default:
		return FALSE
	}
}

func evalBinaryExpression(
	operator token.TokenType,
	left, right object.Object,
) object.Object {
	switch {
	case left.Type() == object.INTEGER && right.Type() == object.INTEGER:
		return evalIntegerBinaryExpression(operator, left, right)

	case left.Type() == object.STRING && right.Type() == object.STRING:
		return evalStringBinaryExpression(operator, left, right)

	case operator == "==":
		return boolToBooleanObject(left == right)
	case operator == "!=":
		return boolToBooleanObject(left != right)

	case left.Type() != right.Type():
		return newError(
			"type mismatch: %s %s %s",
			left.Type(),
			operator,
			right.Type(),
		)
	default:
		return newError(
			"unknown operation: %s %s %s",
			left.Type(),
			operator,
			right.Type(),
		)
	}
}

func evalIntegerBinaryExpression(
	operator token.TokenType,
	left, right object.Object,
) object.Object {
	leftVal := left.(*object.Integer).Value
	rightVal := right.(*object.Integer).Value

	switch operator {
	case token.PLUS:
		return &object.Integer{Value: leftVal + rightVal}
	case token.MINUS:
		return &object.Integer{Value: leftVal - rightVal}
	case token.ASTERISK:
		return &object.Integer{Value: leftVal * rightVal}
	case token.SLASH:
		return &object.Integer{Value: leftVal / rightVal}

	case token.LESS_THAN:
		return boolToBooleanObject(leftVal < rightVal)
	case token.GREATER_THAN:
		return boolToBooleanObject(leftVal > rightVal)
	case token.EQUALS:
		return boolToBooleanObject(leftVal == rightVal)
	case token.NOT_EQUALS:
		return boolToBooleanObject(leftVal != rightVal)

	default:
		return newError(
			"unknown operation: %s %s %s",
			left.Type(),
			operator,
			right.Type(),
		)
	}
}

func evalStringBinaryExpression(
	operator token.TokenType,
	left, right object.Object,
) object.Object {
	if operator != token.PLUS {
		return newError(
			"unknown operation: %s %s %s",
			left.Type(),
			operator,
			right.Type(),
		)
	}

	leftVal := left.(*object.String).Value
	rightVal := right.(*object.String).Value

	return &object.String{Value: leftVal + rightVal}
}

// literals

func evalIdentifier(
	node *ast.Identifier,
	env *object.Environment,
) object.Object {
	if val, ok := env.Get(node.Value); ok {
		return val
	}

	if builtin, ok := builtins[node.Value]; ok {
		return builtin
	}

	return newError("identifier not found: " + node.Value)
}

func evalHashLiteral(
	node *ast.HashLiteral,
	env *object.Environment,
) object.Object {
	pairs := make(map[object.HashKey]object.HashPair)

	for keyNode, valueNode := range node.Pairs {
		key := Eval(keyNode, env)
		if isError(key) {
			return key
		}

		hashKey, ok := key.(object.Hashable)
		if !ok {
			return newError("unusable as hash key: %s", key.Type())
		}

		value := Eval(valueNode, env)
		if isError(value) {
			return value
		}

		hashed := hashKey.HashKey()
		pairs[hashed] = object.HashPair{Key: key, Value: value}
	}

	return &object.Hash{Pairs: pairs}
}

// function

func applyFunction(function object.Object, args []object.Object) object.Object {
	switch function := function.(type) {

	case *object.Function:
		extendedEnv := extendFunctionEnv(function, args)
		evaluated := Eval(function.Body, extendedEnv)
		return unwrapReturnValue(evaluated)

	case *object.Builtin:
		return function.Fn(args...)

	default:
		return newError("not a function: %s", function.Type())
	}
}

func extendFunctionEnv(
	fn *object.Function,
	args []object.Object,
) *object.Environment {
	env := object.NewEnclosedEnvironment(fn.Env)

	for i, param := range fn.Parameters {
		env.Set(param.Value, args[i])
	}

	return env
}

func unwrapReturnValue(obj object.Object) object.Object {
	if val, ok := obj.(*object.ReturnValue); ok {
		return val.Value
	}

	return obj
}

// helpers

func boolToBooleanObject(input bool) *object.Boolean {
	if input {
		return TRUE
	}
	return FALSE
}

func isTruthy(obj object.Object) bool {
	switch obj {
	case NULL:
		return false
	case TRUE:
		return true
	case FALSE:
		return false
	default:
		return true
	}
}

func isError(obj object.Object) bool {
	if obj != nil {
		return obj.Type() == object.ERROR
	}

	return false
}

func newError(format string, a ...interface{}) *object.Error {
	return &object.Error{Message: fmt.Sprintf(format, a...)}
}

File: src/evaluator.zig
const std = @import("std");
const ast = @import("./ast.zig");
const object = @import("./object.zig");
const token = @import("./token.zig");
const builtin = @import("./builtin.zig");
const errors = @import("./errors.zig");

pub const Evaluator = struct {
    allocator: std.mem.Allocator,

    const Self = @This();

    pub fn init(allocator: std.mem.Allocator) Evaluator {
        return .{ .allocator = allocator };
    }

    pub fn eval(
        self: *Self,
        node: ast.Node,
        env: *object.Environment,
    ) Error!object.Object {
        const value_string = try node.print(self.allocator);
        defer self.allocator.free(value_string);
        std.debug.print("EVAL {s} - {s}\n", .{ @tagName(node), value_string });
        return switch (node) {
            .block => |obj| try self.evalBlock(obj.*, env),

            //statements

            .@"return" => |obj| blk: {
                var o = try self.eval(obj.return_value, env);
                if (isError(o)) {
                    return o;
                }
                break :blk .{ .return_value = &o };
            },

            .declaration => |obj| blk: {
                const val = try self.eval(obj.value, env);
                if (isError(val)) {
                    return val;
                }

                _ = env.set(obj.name.value, val);

                break :blk .null;
            },

            // expressions

            .index => |obj| blk: {
                const left = try self.eval(obj.left, env);
                if (isError(left)) {
                    return left;
                }
                const index = try self.eval(obj.index, env);
                if (isError(index)) {
                    return index;
                }

                break :blk try self.evalIndexExpression(
                    left,
                    index,
                    obj.token,
                );
            },

            .call => |obj| blk: {
                const function = try self.eval(obj.function, env);
                if (isError(function)) {
                    return function;
                }
                const args = try self.evalExpressions(
                    obj.arguments,
                    env,
                );
                if (args.len == 1 and isError(args[0])) {
                    return args[0];
                }

                defer self.allocator.free(args);

                break :blk try self.applyFunction(function, args, obj.token);
            },

            .@"if" => |obj| self.evalIfExpression(obj, env),

            .unary_operation => |obj| blk: {
                const right = try self.eval(obj.right, env);
                if (isError(right)) {
                    return right;
                }

                break :blk try self.evalUnaryExpression(
                    obj.operator,
                    right,
                );
            },

            .binary_operation => |obj| blk: {
                const left = try self.eval(obj.left, env);
                if (isError(left)) {
                    return left;
                }
                const right = try self.eval(obj.right, env);
                if (isError(right)) {
                    return right;
                }

                break :blk try self.evalBinaryExpression(
                    obj.operator,
                    left,
                    right,
                );
            },

            // literals

            .identifier => |obj| self.evalIdentifier(obj, env),

            .boolean_literal => |obj| .{ .boolean = obj.value },

            .integer_literal => |obj| blk: {
                break :blk .{ .integer = obj.value };
            },

            .string_literal => |obj| blk: {
                break :blk .{ .string = obj.value };
            },

            .array_literal => |obj| blk: {
                const elements = try self.evalExpressions(
                    obj.elements,
                    env,
                );
                const val = try object.Array.init(
                    self.allocator,
                    elements,
                );
                break :blk .{ .array = val };
            },

            .hash_literal => |obj| try self.evalHashLiteral(
                obj,
                env,
            ),

            .function_literal => |obj| blk: {
                const val = try object.Function.init(
                    self.allocator,
                    obj.parameters,
                    obj.body,
                    env,
                );
                break :blk .{ .function = val };
            },
        };
    }

    fn evalBlock(
        self: *Self,
        node: ast.Block,
        env: *object.Environment,
    ) !object.Object {
        var result: object.Object = undefined;

        for (node.statements) |stmt| {
            result = try self.eval(stmt, env);

            _ = switch (result) {
                .return_value => return result,
                .@"error" => return result,
                else => void,
            };
        }

        return result;
    }

    // expressions

    fn evalExpressions(
        self: *Self,
        nodes: []ast.Node,
        env: *object.Environment,
    ) ![]object.Object {
        var result = std.ArrayList(object.Object).init(
            self.allocator,
        );

        for (nodes) |n| {
            const val = try self.eval(n, env);
            try result.append(val);
            if (isError(val)) {
                return result.items[result.items.len - 1 .. result.items.len];
            }
        }

        return try result.toOwnedSlice();
    }

    fn evalIndexExpression(
        self: *Self,
        left: object.Object,
        index: object.Object,
        tok: token.Token,
    ) !object.Object {
        return switch (left) {
            .array => |obj| self.evalArrayIndexExpression(
                obj.*,
                index,
            ),
            .hash => |obj| try self.evalHashIndexExpression(
                obj.*,
                index,
                tok,
            ),
            else => try self.newError(
                "index operator is not supported on {s}",
                .{left.stringify(self.allocator)},
                tok,
            ),
        };
    }

    fn evalArrayIndexExpression(
        _: *Self,
        array: object.Array,
        index: object.Object,
    ) object.Object {
        const idx = index.integer;
        const max = array.elements.len - 1;

        if (idx < 0 or idx > max) {
            return .null;
        }

        return array.elements[@intCast(idx)];
    }

    fn evalHashIndexExpression(
        self: *Self,
        hash: object.Hash,
        index: object.Object,
        tok: token.Token,
    ) !object.Object {
        const hash_key = index.hashKey();

        if (hash_key == null) {
            return try self.newError(
                "{s} is unusable as hash key",
                .{index.stringify(self.allocator)},
                tok,
            );
        }

        const pair = hash.pairs.get(hash_key.?);

        if (pair == null) {
            return .null;
        }

        return pair.?.value;
    }

    fn evalIfExpression(
        self: *Self,
        node: *ast.If,
        env: *object.Environment,
    ) !object.Object {
        const condition = try self.eval(node.condition, env);
        if (isError(condition)) {
            return condition;
        }

        if (isTruthy(condition)) {
            return self.eval(.{ .block = node.consequence }, env);
        } else if (node.alternative) |alt| {
            return self.eval(.{ .block = alt }, env);
        } else {
            return .null;
        }
    }

    fn evalUnaryExpression(
        self: *Self,
        operator: token.Token,
        right: object.Object,
    ) !object.Object {
        const obj: ?object.Object = switch (operator.type) {
            .minus => try self.evalNegateExpression(right),
            .bang => self.evalBooleanNotExpression(right),
            else => null,
        };

        if (obj) |o| {
            return o;
        }

        return try self.newError(
            "unknown operation: {s}{s}",
            .{ operator.literal, right.stringify(self.allocator) },
            operator,
        );
    }

    fn evalNegateExpression(
        _: *Self,
        right: object.Object,
    ) !?object.Object {
        return switch (right) {
            .integer => |obj| blk: {
                break :blk .{ .integer = -obj };
            },
            else => null,
        };
    }

    fn evalBooleanNotExpression(
        _: *Self,
        right: object.Object,
    ) object.Object {
        return .{ .boolean = isTruthy(right) };
    }

    fn evalBinaryExpression(
        self: *Self,
        operator: token.Token,
        left: object.Object,
        right: object.Object,
    ) !object.Object {
        var obj: ?object.Object = null;

        if (left == object.ObjectType.integer and right == object.ObjectType.integer) {
            obj = try self.evalIntegerBinaryExpression(
                operator,
                left.integer,
                right.integer,
            );
        }

        if (left == object.ObjectType.string and right == object.ObjectType.string) {
            obj = try self.evalStringBinaryExpression(
                operator,
                left.string,
                right.string,
            );
        }

        if (obj) |o| {
            return o;
        }

        if (operator.type == token.TokenType.equals) {
            return .{ .boolean = std.meta.eql(left, right) };
        }
        if (operator.type == token.TokenType.not_equals) {
            return .{ .boolean = !std.meta.eql(left, right) };
        }
        if (@intFromEnum(left) != @intFromEnum(right)) {
            return try self.newError(
                "type mismatch: {s} {s} {s}",
                .{
                    left.stringify(self.allocator),
                    operator.literal,
                    right.stringify(self.allocator),
                },
                operator,
            );
        }
        return try self.newError(
            "unknown operation: {s} {s} {s}",
            .{
                left.stringify(self.allocator),
                operator.literal,
                right.stringify(self.allocator),
            },
            operator,
        );
    }

    fn evalIntegerBinaryExpression(
        _: *Self,
        operator: token.Token,
        left: object.Integer,
        right: object.Integer,
    ) !?object.Object {
        return switch (operator.type) {
            .plus => blk: {
                break :blk .{ .integer = left + right };
            },
            .minus => blk: {
                break :blk .{ .integer = left - right };
            },
            .asterisk => blk: {
                break :blk .{ .integer = left * right };
            },
            .slash => blk: {
                break :blk .{ .integer = @divTrunc(left, right) };
            },

            .less_than => .{ .boolean = left < right },
            .greater_than => .{ .boolean = left > right },
            .equals => .{ .boolean = left == right },
            .not_equals => .{ .boolean = left != right },

            else => null,
        };
    }

    fn evalStringBinaryExpression(
        self: *Self,
        operator: token.Token,
        left: object.String,
        right: object.String,
    ) !?object.Object {
        return switch (operator.type) {
            .plus => blk: {
                const str = try std.fmt.allocPrint(
                    self.allocator,
                    "{s}{s}",
                    .{ left, right },
                );
                break :blk .{ .string = str };
            },

            else => null,
        };
    }

    // literals

    fn evalIdentifier(
        self: *Self,
        node: *ast.Identifier,
        env: *object.Environment,
    ) !object.Object {
        const val = env.get(node.value);

        if (val) |v| {
            return v;
        }

        const builtin_fn = builtin.builtins.get(node.value);
        if (builtin_fn) |b| {
            const v = try object.Builtin.init(self.allocator, b);
            return .{ .builtin = v };
        }

        return try self.newError(
            "identifier '{s}' not found",
            .{node.value},
            node.token,
        );
    }

    fn evalHashLiteral(
        self: *Self,
        node: *ast.HashLiteral,
        env: *object.Environment,
    ) !object.Object {
        var pairs = std.AutoHashMap(
            object.HashKey,
            *object.HashPair,
        ).init(self.allocator);

        var iterator = node.pairs.iterator();
        while (iterator.next()) |pair| {
            const key = try self.eval(pair.key_ptr.*, env);
            if (isError(key)) {
                return key;
            }

            const hash_key = key.hashKey();
            if (hash_key == null) {
                return try self.newError(
                    "{s} is unusable as hash key",
                    .{key.stringify(self.allocator)},
                    node.token,
                );
            }

            const value = try self.eval(pair.value_ptr.*, env);
            if (isError(value)) {
                return value;
            }

            const hash_pair = try object.HashPair.init(
                self.allocator,
                key,
                value,
            );
            _ = try pairs.put(hash_key.?, hash_pair);
        }

        const val = try object.Hash.init(self.allocator, pairs);
        return .{ .hash = val };
    }

    // function

    fn applyFunction(
        self: *Self,
        function: object.Object,
        args: []object.Object,
        tok: token.Token,
    ) !object.Object {
        return switch (function) {
            .function => |obj| blk: {
                if (obj.parameters.len != args.len) {
                    break :blk try self.invalidArgumentAmountError(
                        obj.parameters.len,
                        args.len,
                        tok,
                    );
                }

                for (args) |arg| {
                    arg.ref();
                }

                defer {
                    for (args) |arg| {
                        _ = arg.deref(self.allocator);
                    }
                }

                const extended_env = try self.extendFunctionEnv(
                    obj.*,
                    args,
                );

                const val = try self.eval(
                    .{ .block = obj.body },
                    extended_env,
                );

                defer extended_env.deref();
                break :blk self.unwrapReturnValue(val);
            },

            .builtin => |obj| blk: {
                defer _ = function.deref(self.allocator);
                break :blk obj.function(self, args, tok);
            },

            else => blk: {
                const fn_string = function.stringify(self.allocator);
                defer self.allocator.free(fn_string);
                break :blk try self.newError(
                    "{s} is not a function",
                    .{fn_string},
                    tok,
                );
            },
        };
    }

    fn extendFunctionEnv(
        self: *Self,
        function: object.Function,
        args: []object.Object,
    ) !*object.Environment {
        var env = try object.Environment.initEnclosed(
            self.allocator,
            function.env,
        );

        for (function.parameters, 0..) |param, i| {
            _ = env.set(param.value, args[i]);
        }

        return env;
    }

    fn unwrapReturnValue(
        _: *Self,
        obj: object.Object,
    ) object.Object {
        return switch (obj) {
            .return_value => |o| o.*,
            else => obj,
        };
    }

    // error

    pub fn newError(
        self: *Self,
        comptime format: []const u8,
        a: anytype,
        tok: token.Token,
    ) !object.Object {
        const msg = try std.fmt.allocPrint(
            self.allocator,
            format,
            a,
        );
        const err = try object.Error.init(
            self.allocator,
            msg,
            tok,
        );
        return .{ .@"error" = err };
    }

    pub fn invalidArgumentAmountError(
        self: *Self,
        expect: usize,
        got: usize,
        tok: token.Token,
    ) !object.Object {
        return try self.newError(
            "expected {d} argument{s}, got {d}",
            .{
                expect,
                if (expect % 10 == 1) "" else "s",
                got,
            },
            tok,
        );
    }

    // helpers

    fn isError(obj: object.Object) bool {
        return switch (obj) {
            .@"error" => true,
            else => false,
        };
    }

    fn isTruthy(obj: object.Object) bool {
        return switch (obj) {
            .boolean => |o| o,
            .null => false,
            else => true,
        };
    }

    pub const Error = error{
        OutOfMemory,
        InputOutput,
        SystemResources,
        OperationAborted,
        BrokenPipe,
        ConnectionResetByPeer,
        WouldBlock,
        AccessDenied,
        Unexpected,
        DiskQuota,
        FileTooBig,
        NoSpaceLeft,
        DeviceBusy,
        InvalidArgument,
        NotOpenForWriting,
        LockViolation,
    };
};

test {
    _ = @import("./tests/evaluator_test.zig");
}
